python/cuml/fil/fil.pyx

```diff
@@ -0,0 +1,340 @@
+        self.handle = new_handle
+
+    cdef ModelHandle get_handle(self):
+        return self.handle
+
+    def __cdel__(self):
+        if self.handle != NULL:
+            TreeliteFreeModel(self.handle)
+
+    @property
+    def num_trees(self):
+        assert self.handle != NULL
+        cdef size_t out
+        TreeliteQueryNumTree(self.handle, &out)
+        return out
+
+    @property
+    def num_features(self):
+        assert self.handle != NULL
+        cdef size_t out
+        TreeliteQueryNumFeature(self.handle, &out)
+        return out
+
+    @staticmethod
+    def from_filename(filename, model_type="xgboost"):
+        """
+        Returns a TreeliteModel object loaded from `filename`
+
+        Parameters
+        ----------
+        filename : string
+            Path to treelite model file to load
+
+        model_type : string
+            Type of model: 'xgboost', 'protobuf', or 'lightgbm'
+        """
+        filename_bytes = filename.encode("UTF-8")
+        cdef ModelHandle handle
+        if model_type == "xgboost":
+            res = TreeliteLoadXGBoostModel(filename_bytes, &handle)
+            if res < 0:
+                raise RuntimeError("Failed to load %s" % filename)
+        elif model_type == "protobuf":
+            # XXX Not tested
+            res = TreeliteLoadProtobufModel(filename_bytes, &handle)
+            if res < 0:
+                raise RuntimeError("Failed to load %s" % filename)
+        elif model_type == "lightgbm":
+            res = TreeliteLoadLightGBMModel(filename_bytes, &handle)
+            if res < 0:
+                raise RuntimeError("Failed to load %s" % filename)
+        else:
+            raise ValueError("Unknown model type %s" % model_type)
+        model = TreeliteModel()
+        model.set_handle(handle)
+        return model
+
+
+cdef extern from "fil/fil.h" namespace "ML::fil":
+    cdef enum algo_t:
+        NAIVE,
+        TREE_REORG,
+        BATCH_TREE_REORG
+
+    cdef struct forest:
+        pass
+
+    ctypedef forest* forest_t
+
+    cdef struct treelite_params_t:
+        algo_t algo
+        bool output_class
+        float threshold
+
+    cdef void free(cumlHandle& handle,
+                   forest_t)
+
+    cdef void predict(cumlHandle& handle,
+                      forest_t,
+                      float*,
+                      float*,
+                      size_t)
+
+    cdef forest_t from_treelite(cumlHandle& handle,
+                                forest_t*,
+                                ModelHandle,
+                                treelite_params_t*)
+
+cdef class ForestInference_impl():
+
+    cpdef object handle
+    cdef forest_t forest_data
+
+    def __cinit__(self,
+                  handle=None):
+        self.handle = handle
+
+    def get_algo(self, algo_str):
+        if algo_str == 'NAIVE':
+            return algo_t.NAIVE
+        elif algo_str == 'BATCH_TREE_REORG':
+            return algo_t.BATCH_TREE_REORG
+        else:
```

Comment:
`elif algo_str == 'TREE_REORG':` and throw an exception in the `else` branch. We want to detect invalid parameters rather than silently accepting them.

A better alternative is to replace the conditional by a map lookup.